File: sdk/go/pwrap/rest.go
package pwrap

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// RestToken is what /v1/rest/token returns. The client hits URL with `Authorization: Bearer Token`
// against PostgREST. Refresh before ExpiresAt.
type RestToken struct {
	Token     string    `json:"token"`
	URL       string    `json:"url"`
	Role      string    `json:"role"`
	ExpiresAt time.Time `json:"expires_at"`
}

// RestTokenOpts lets callers attach an app-level user id (for RLS policies referencing
// `request.jwt.claims->>'user_id'`) and override the default TTL.
type RestTokenOpts struct {
	UserID     string
	TTLSeconds int
}

// IssueRestToken exchanges the cached API key for a PostgREST-compatible JWT.
// Returns ErrRestDisabled if pwrapd was configured without REST support.
func (c *Client) IssueRestToken(ctx context.Context, opts RestTokenOpts) (RestToken, error) {
	base := strings.TrimRight(c.cfg.ControlURL, "/")
	body, err := json.Marshal(map[string]any{
		"user_id":     opts.UserID,
		"ttl_seconds": opts.TTLSeconds,
	})
	if err != nil {
		return RestToken{}, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, base+"/v1/rest/token", bytes.NewReader(body))
	if err != nil {
		return RestToken{}, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.cfg.APIKey)

	resp, err := c.http.Do(req)
	if err != nil {
		return RestToken{}, fmt.Errorf("pwrap: rest token: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusServiceUnavailable {
		return RestToken{}, ErrRestDisabled
	}
	if resp.StatusCode >= 400 {
		b, _ := io.ReadAll(resp.Body)
		return RestToken{}, fmt.Errorf("pwrap: /v1/rest/token: %s %s", resp.Status, bytes.TrimSpace(b))
	}
	var out RestToken
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return RestToken{}, err
	}
	return out, nil
}

// ErrRestDisabled is returned when pwrapd reports REST support is off.
var ErrRestDisabled = errors.New("pwrap: REST integration not enabled on this pwrapd")

File: sdk/py/pwrap/rest.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Optional

from .exceptions import PwrapError, RestDisabledError


@dataclass
class RestToken:
    token: str
    url: str
    role: str
    expires_at: datetime


async def issue_rest_token(
    client,
    *,
    user_id: Optional[str] = None,
    ttl_seconds: Optional[int] = None,
) -> RestToken:
    """Exchange the SDK's API key for a PostgREST-compatible JWT."""
    body: dict = {}
    if user_id:
        body["user_id"] = user_id
    if ttl_seconds:
        body["ttl_seconds"] = ttl_seconds
    resp = await client._http_.post(
        client._control_url_ + "/v1/rest/token",
        headers={
            "Authorization": f"Bearer {client._api_key_}",
            "Content-Type": "application/json",
        },
        json=body,
    )
    if resp.status_code == 503:
        raise RestDisabledError("pwrapd has no JWT secret configured")
    if resp.status_code >= 400:
        raise PwrapError(f"/v1/rest/token {resp.status_code}: {resp.text.strip()}")
    data = resp.json()
    return RestToken(
        token=data["token"],
        url=data["url"],
        role=data["role"],
        expires_at=datetime.fromisoformat(data["expires_at"].replace("Z", "+00:00")),
    )
